File: src/registry.ts
// Session registry: one JSON file per live Claude Code session, keyed by
// session id, under ~/.claude-voice/registry/. Shared contract lives in the
// ait Phase 1 epic (handy-UkLWZ.2).
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export interface RegistryEntry {
  session_id: string;
  claude_pid: number;
  server_pid: number;
  port: number;
  cwd: string;
  started_at: string;
  name: string | null;
}

export function defaultBaseDir(): string {
  return join(homedir(), ".claude-voice");
}

export function registryDir(base: string = defaultBaseDir()): string {
  return join(base, "registry");
}

export function ensureDirs(base: string = defaultBaseDir()): void {
  mkdirSync(registryDir(base), { recursive: true, mode: 0o700 });
}

export function entryPath(
  sessionId: string,
  base: string = defaultBaseDir(),
): string {
  return join(registryDir(base), `${sessionId}.json`);
}

export function writeEntry(
  entry: RegistryEntry,
  base: string = defaultBaseDir(),
): void {
  ensureDirs(base);
  writeFileSync(
    entryPath(entry.session_id, base),
    `${JSON.stringify(entry, null, 2)}\n`,
  );
}

export function readEntry(
  sessionId: string,
  base: string = defaultBaseDir(),
): RegistryEntry | null {
  try {
    return JSON.parse(
      readFileSync(entryPath(sessionId, base), "utf8"),
    ) as RegistryEntry;
  } catch {
    return null;
  }
}

export function removeEntry(
  sessionId: string,
  base: string = defaultBaseDir(),
): void {
  rmSync(entryPath(sessionId, base), { force: true });
}

/** Identity handling: decide this server's registry entry from its spawn
 *  environment. Returns null when there is no owning session to register
 *  against (e.g. run standalone for a smoke test). */
export function buildSelfEntry(input: {
  env: Record<string, string | undefined>;
  port: number;
  cwd: string;
  serverPid: number;
  claudePid: number;
}): RegistryEntry | null {
  const sessionId = input.env.CLAUDE_CODE_SESSION_ID;
  if (!sessionId) return null;
  return {
    session_id: sessionId,
    claude_pid: input.claudePid,
    server_pid: input.serverPid,
    port: input.port,
    cwd: input.cwd,
    started_at: new Date().toISOString(),
    name: null, // reserved for Phase 2 named sessions
  };
}

export function pidAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

/** Remove entries whose owning claude process is gone (or that are unreadable).
 *  Returns the removed filenames. `isAlive` is injectable for tests. */
export function sweep(
  base: string = defaultBaseDir(),
  isAlive: (pid: number) => boolean = pidAlive,
): string[] {
  const dir = registryDir(base);
  if (!existsSync(dir)) return [];
  const removed: string[] = [];
  for (const file of readdirSync(dir)) {
    if (!file.endsWith(".json")) continue;
    const path = join(dir, file);
    let dead = true;
    try {
      const entry = JSON.parse(readFileSync(path, "utf8")) as RegistryEntry;
      dead = !Number.isInteger(entry.claude_pid) || !isAlive(entry.claude_pid);
    } catch {
      // unreadable entry: treat as dead
    }
    if (dead) {
      rmSync(path, { force: true });
      removed.push(file);
    }
  }
  return removed;
}

File: src/registry.py
"""Session registry: one JSON file per live Claude Code session, keyed by
session id, under ~/.claude-voice/registry/. Shared contract lives in the
ait Phase 1 epic (handy-UkLWZ.2); Python port of registry.ts (behaviour spec:
registry_test.py, ported from registry.test.ts)."""

import json
import os
from pathlib import Path


def default_base_dir() -> Path:
    return Path.home() / ".claude-voice"


def registry_dir(base: Path | None = None) -> Path:
    return (base or default_base_dir()) / "registry"


def ensure_dirs(base: Path | None = None) -> None:
    registry_dir(base).mkdir(parents=True, exist_ok=True, mode=0o700)


def entry_path(session_id: str, base: Path | None = None) -> Path:
    return registry_dir(base) / f"{session_id}.json"


def write_entry(entry: dict, base: Path | None = None) -> None:
    ensure_dirs(base)
    entry_path(entry["session_id"], base).write_text(json.dumps(entry, indent=2) + "\n")


def read_entry(session_id: str, base: Path | None = None) -> dict | None:
    try:
        return json.loads(entry_path(session_id, base).read_text())
    except Exception:
        return None


def remove_entry(session_id: str, base: Path | None = None) -> None:
    entry_path(session_id, base).unlink(missing_ok=True)


def remove_own_entry(session_id: str, server_pid: int, base: Path | None = None) -> bool:
    """Remove the entry only if this server still owns it.

    Claude Code spawns the server twice at session start (observed 2026-07-23,
    ant handy-C8ggs); the first instance's shutdown must not delete the entry
    the second instance just wrote. Returns True if an entry was removed.
    """
    current = read_entry(session_id, base)
    if current is None or current.get("server_pid") != server_pid:
        return False
    remove_entry(session_id, base)
    return True


def build_self_entry(
    env: dict, port: int, cwd: str, server_pid: int, claude_pid: int
) -> dict | None:
    """Identity handling: decide this server's registry entry from its spawn
    environment. Returns None when there is no owning session to register
    against (e.g. run standalone for a smoke test)."""
    session_id = env.get("CLAUDE_CODE_SESSION_ID")
    if not session_id:
        return None
    from datetime import datetime, timezone

    return {
        "session_id": session_id,
        "claude_pid": claude_pid,
        "server_pid": server_pid,
        "port": port,
        "cwd": cwd,
        "started_at": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "name": None,  # reserved for Phase 2 named sessions
    }


def pid_alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True


def sweep(base: Path | None = None, is_alive=pid_alive) -> list[str]:
    """Remove entries whose owning claude process is gone (or that are
    unreadable). Returns the removed filenames. `is_alive` is injectable for
    tests."""
    directory = registry_dir(base)
    if not directory.is_dir():
        return []
    removed: list[str] = []
    for path in sorted(directory.iterdir()):
        if path.suffix != ".json":
            continue
        dead = True
        try:
            entry = json.loads(path.read_text())
            pid = entry.get("claude_pid")
            dead = not isinstance(pid, int) or not is_alive(pid)
        except Exception:
            pass  # unreadable entry: treat as dead
        if dead:
            path.unlink(missing_ok=True)
            removed.append(path.name)
    return removed
